server/server.go

```diff
@@ -132,12 +132,13 @@ func (srv *Server) Stop() {
 	// stop accept new connections
 	srv.listener.Close()
 
-	// TODO fatal error: concurrent map iteration and map write
 	var wg sync.WaitGroup
+	srv.connLock.Lock()
```

Comment:
It is naive solution, cause conn.safeClose has server.removeConnection where it has lock too that cause deadlock. In this cause it is not so trivial to realize safe and stable graceful shutdouwn

Reply:
Fair point, that does not cause a deadlock though as all the go routines will wait until the lock is released to start closing themselves, since the lock is released once all goroutines are initialized, this should not be an issue (at least I have not experienced any in our use cases)

Reply:
Hm... will experiment your solution and come back with feedback

Reply:
All is ok